File: Proga/Lab8/MyClass/MyList.cs
using System;
using System.Collections.Generic;
using System.Collections;

namespace MyClasses
{
    public class MyList
    {
        private Node _head;
        private Node _tail;
        public MyList()
        {
            _head = null;
            _tail = null;
        }
        public void put(long num)
        {
            Node temp = new Node(num);
            this.AddLast(temp);
        }
        public uint Length()
        {
            uint counter = 0;
            Node q = _head;
            while (q.next != null)
            {
                counter++;
                q = q.next;
            }
            return counter;
        }
        public uint CountElemDivBy5OnEvenPos()
        {
            bool IsEven = false;
            uint counter = 0;
            Node q = _head;
            while (q.next != null)
            {
                if (IsEven)
                    if (q.x % 5 == 0)
                        counter++;
                IsEven = !IsEven;
                q = q.next;
            }
            return counter;
        }
        public void RemoveElementsMoreThanAverage()
        {
            Node q = _head;
            long averageLower = CountAverageLower();
            while (q != null)
            {
                Node p = q.next;
                if (q.x > averageLower)
                    DeleteNode(q);
                q = p;
            }
        }
        public long CountAverageLower()
        {
            long sum = 0;
            uint counter = 0;
            Node q = _head;
            while (q != null)
            {
                sum += q.x;
                counter++;
                q = q.next;
            }
            return sum / counter;
        }
        public override string ToString()
        {
            string strOut = "";
            Node q = _head;
            while (q != null)
            {
                strOut += (q.x).ToString();
                if (q.next != null)
                    strOut += ", ";
                else
                    strOut += ".";
                q = q.next;
            }
            return strOut;
        }
        private void AddLast(Node newNode)
        {
            if (_head == null)
            {
                _head = _tail = newNode;
            }
            else
            {
                newNode.prev = _tail;
                _tail.next = newNode;
                _tail = newNode;
            }
        }
        private void DeleteNode(Node p)
        {
            Node q = _head;
            if (_head == p)
            {
                _head = p.next;
            }
            else
            {
                while (q != null && q.next != p)
                {
                    q = q.next;
                }
                if (q != null)
                    q.next = p.next;
            }
        }
    }
}

File: Proga/Lab8/Lab8/MyList.cpp
#include <iostream>
#include "MyList.h"

MyList::MyList()
{
    _head = NULL;
    _tail = NULL;
}

void MyList::put(long num)
{
    PNode temp = CreateNode(num);
    this->AddLast(temp);
}

size_t MyList::Length()
{
    size_t counter = 0;
    PNode q = _head;
    while (q->next)
    {
        counter++;
        q = q->next;
    }
    return counter;
}

std::ostream& operator<< (std::ostream& os, MyList& listOut)
{
    PNode q = listOut._head;
    while (q)
    {
        os << q->x;
        if (q->next)
            os << ", ";
        else
            os << ".";
        q = q->next;
    }
    return os;
}

size_t MyList::CountElemDivBy5OnEvenPos()
{
    bool IsEven = false;
    size_t counter = 0;
    PNode q = _head;
    while (q->next)
    {
        if (IsEven)
            if (q->x % 5 == 0)
                counter++;
        IsEven = !IsEven;
        q = q->next;
    }
    return counter;
}

void MyList::RemoveElementsMoreThanAverage()
{
    PNode q = _head;
    long averageLower = CountAverageLower();
    while (q)
    {
        PNode p = q->next;
        if (q->x > averageLower)
            DeleteNode(q);
        q = p;
    }
}

long MyList::CountAverageLower()
{
    long sum = 0;
    size_t counter = 0;
    PNode q = _head;
    while (q)
    {
        sum += q->x;
        counter++;
        q = q->next;
    }
    return sum / counter;
}

MyList::~MyList()
{
    while (_head)
    {
        _tail = _head->next;
        delete _head;
        _head = _tail;
    }
}

PNode MyList::CreateNode(long num)
{
    PNode newNode = new Node;
    newNode->x = num;
    newNode->next = NULL;
    newNode->prev = NULL;
    return newNode;
}

void MyList::AddLast(PNode newNode)
{
    if (_head == NULL)
    {
        _head = _tail = newNode;
    }
    else
    {
        newNode->prev = _tail;
        _tail->next = newNode;
        _tail = newNode;
    }
}

void MyList::DeleteNode(PNode p)
{
    PNode q = _head;
    if (_head == p)
    {
        _head = p->next;
    }
    else
    {
        while (q && q->next != p)
        {
            q = q->next;
        }
        if (q != NULL)
            q->next = p->next;
    }
    delete p;
}